### ISSUE TITLE
[Question] Is it possible to enqueue jobs outside a batch while inside some batch?

### ISSUE THREAD

**Issue by Sauloxd** [mperham/sidekiq#5200]
Ruby version: 2.7.5
Sidekiq / Pro / Enterprise version(s): Sidekiq 6.4.1 / Pro 5.3.1 / Ent - not yet but soon

We are building a complex batch workflow, trying to improve a legacy flow that now needs scaling, using [this guide](https://github.com/mperham/sidekiq/wiki/Really-Complex-Workflows-with-Batches) as reference.

Basically, we are trying to mass create records (about ~100k) in parallel but we have some legacy ActiveRecord::Callbacks (not to be confused with Sidekiq::Batch::Callback) that are triggered after model creation... enqueuing another job.

This job is now part of the batch, and can cause the batch to halt if it raises an error and exhaust retries and such.
For us, this job that could be executed way later (we run for data consistency), and don't really need to be part of the batch

Is there a way to enqueue a job without an associated BID, even though this job was enqueued FROM a job inside a batch?
Is this even a good practice? to enqueue jobs from a batch without associating it to the batch?

Example of flow:

1. OverallBatch
   1. Step 1:
      1. Push bulk 100k job that will verify all lines in CSV
      2. Each individual job verify if line is valid
   2. Step 2:
      1. Push bulk 100k job that will create all validated lines
      2. Each individual job creates the Entity
         1. Each creation triggers the SyncSomethingJob in :after_create callbacks in this Entity
         2. This generates 100k roundtrips to redis
         3. Sadly these jobs fails sometimes, and it causes the whole batch to halt

**Comment by mperham** [comment]
Any job enqueued within a `batch.jobs` block will be part of that batch. OTOH, a job which is created inside a batch job BUT is not within a `batch.jobs` block will not be part of the batch.

It sounds like the failure is in pushing the SyncSomethingJob to Redis? If you don't want errors within the individual jobs created in 1.ii.a to halt Step 2, you can create them outside of a batch or use a :complete callback instead of :success. Otherwise make sure you've enabled retries and gradually improve the app code to handle any app-specific edge cases which are causing the stoppage. Sometimes this can take time if processing "dirty" external data.

**Comment by Sauloxd** [comment]
Thanks for the fast response!
You are absolutely right, I think I interpreted wrong the logs, because I saw the BID in it, like:
```
pid=22 tid=zv67u class=CreateEntityJob jid=1234 bid=1234 [INFO] SidekiqJob enqueued SyncSomethingJob  (Job ID: 1234) to Sidekiq(lowest_priority) with arguments: [1234]
```

So by having bid in its `Sidekiq::Context.current` and in the logs, I really thought it was being associated with it!
After seeing your response, I investigated more, and finally found a superfetch log reporting that one of `CreateEntityJob` (that is part of the batch jobs) was recovered and after that, the batch really succeeded.

Just did some testing here by forcing an error in the `SyncSomethingJob` and it really showed the `bid` when sidekiq client enqueued the job, but it had no influence in the batch whatsoever.

Thanks for your help, and the tips for improving our batch flow!
